util/network-devp2p/src/handshake.rs

```diff
@@ -314,27 +311,26 @@ impl Handshake {
         Message: Send + Clone + Sync + 'static,
     {
         trace!(target: "network", "Sending handshake auth to {:?}", self.connection.remote_addr_str());
-        let mut data = [0u8; /*Signature::SIZE*/ 65 + /*H256::SIZE*/ 32 + /*Public::SIZE*/ 64 + /*H256::SIZE*/ 32 + 1]; //TODO: use associated constants
-        let len = data.len();
-        {
-            data[len - 1] = 0x0;
-            let (sig, rest) = data.split_at_mut(65);
-            let (hepubk, rest) = rest.split_at_mut(32);
-            let (pubk, rest) = rest.split_at_mut(64);
-            let (nonce, _) = rest.split_at_mut(32);
+        let mut rlp = RlpStream::new_list(4);
+        let shared = *ecdh::agree(secret, &self.id)?;
+        rlp.append(&sign(self.ecdhe.secret(), &(shared ^ self.nonce))?.to_vec());
+        rlp.append(public);
+        rlp.append(&self.nonce);
+        rlp.append(&PROTOCOL_VERSION);
 
-            // E(remote-pubk, S(ecdhe-random, ecdh-shared-secret^nonce) || H(ecdhe-random-pubk) || pubk || nonce || 0x0)
-            let shared = *ecdh::agree(secret, &self.id)?;
-            sig.copy_from_slice(&*sign(self.ecdhe.secret(), &(shared ^ self.nonce))?);
-            write_keccak(self.ecdhe.public(), hepubk);
-            pubk.copy_from_slice(public);
-            nonce.copy_from_slice(&self.nonce);
-        }
-        let message = ecies::encrypt(&self.id, &[], &data)?;
-        self.auth_cipher = message.clone();
-        self.connection.send(io, message);
-        self.connection.expect(V4_ACK_PACKET_SIZE);
-        self.state = HandshakeState::ReadingAck;
+        let encoded = rlp
+            .out()
+            .into_iter()
+            .chain(repeat(0).take(rand::thread_rng().gen_range::<usize>(100, 301)))
```

Comment:
Let's avoid premature pessimization and use a static array of zeroes and `Vec::extend` instead.
LLVM is pretty decent, but it's not magic: https://rust.godbolt.org/z/vjobKj.
The latter generates much less code and uses one single memset.

Reply:
Replaced with `Vec::resize`.